### PR TITLE
HZN-1520: Rewrite Geocoder Services and provide Configuration UI

### PR DESCRIPTION
**Pull request by mvrueden**
* JIRA: https://issues.opennms.org/browse/HZN-1520

With this the Nominatim, MapQuest and Google geocoders are now available.
Unfortunately they are all not working out of the box and require some configuration.
You can get to the UI via `Admin -> Configure Geocoder Service`.
Please note, that existing configuration is not compatible and you must definitely configure your Geocoder Service:

The Nominatim Geocoder just requires you to accept the usage terms and provide an email.

The Google Geocoder either requires a API Key or clientId/signature authentication.
An API Key is provided in the Jira issue (only visible by TOG)
I don't have any clientId/signature credentials.

The MapQuest Geocoder requires an API Key.
An API Key is provided in the Jira issue (only visible by TOG)

The following things have been tested:
 - Using the UI to configure each geocoder
 - Activating/Deactivating them (and see if provisioning will work accordingly) (There will be log entries telling you which geocoder service was used (or if disabled or not configured properly)
 - Update configuration via the cfg files and see if they are reflected in the ui (may take some time)

I was not able to test Google's clientId/Signature authentication.

Due to our setup in rare situations you might get an error when changing configuration options and reload manually very quickly. This is caused because any change will refresh the bundles to reload. This means also the rest service bundle. And in some weird situations you might get a 404. This is nothing I can easily prevent, so a message is shown to the user. You can simulate this by simply stopping the Geocoder Rest Bundle and access the ui

Feel free to provide some ideas/feedback and test this out.

**All Geocoders disabled**
Screenshot 2019-04-30 at 17 58 28

**Nominatim Geocoder enabled, but not properly configured**
Screenshot 2019-04-30 at 17 58 34

**Nominatim Geocoder Configuration Page**
Screenshot 2019-04-30 at 17 58 42

**Google Geocoder enabled and properly configured**
Screenshot 2019-04-30 at 17 58 56

### CODE REVIEW

**Comment by mvrueden** [comment]
I am code-wise finished, just want to add some more documentation/comments and maybe rename certain things. I added the `triage` label to indicate this.

**Comment by pioto** [comment]
Looks like this leaves the configuration options in `opennms.properties` intact; if that configuration is no longer valid, it should probably be removed in this PR as well?

https://github.com/OpenNMS/opennms/blob/develop/opennms-base-assembly/src/main/filtered/etc/opennms.properties#L535-L564, plus a few other sections, mention Geocoding there right now.

**Comment by mvrueden** [comment]
Yeah, you are right. It is on my list of left clean up steps, but forgot to mention that above. Sorry (-:

Just as a general note: The properties are not used at all, even without these changes.

File: opennms-webapp/src/main/webapp/admin/geoservice/index.jsp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+--%>
+<jsp:include page="/includes/bootstrap.jsp" flush="false">
+    <jsp:param name="norequirejs" value="true" />
+
+    <jsp:param name="title" value="Flow Classification" />
```
Comment (christianpape) [inline opennms-webapp/src/main/webapp/admin/geoservice/index.jsp]: The title is wrong

File: opennms-webapp/src/main/webapp/admin/geoservice/index.jsp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    <jsp:param name="headTitle" value="Flow Classification" />
```
Comment (christianpape) [inline opennms-webapp/src/main/webapp/admin/geoservice/index.jsp]: same here

**Comment by mvrueden** [comment]
> I've built an instance from the RPMS and noticed a copy&paste error while testing. The tab label is still showing "Flow Classification" instead of "Geocoder Configuration"
> 
> geolocation-tablabel

No matter how hard you try, you always miss something. This is now fixed.

**Comment by mvrueden** [comment]
> Just FYI, not a deal breaker here, I've noticed when using the Google Gecoder service I got first an error message with a timeout:
> 
> ```
> 2019-05-02T11:33:25,272 | DEBUG | pool-11-thread-1 | GoogleGeocoderService            | 231 - org.opennms.features.geocoder.google - 24.1.0.SNAPSHOT | Configuration: {clientId=, apiKey=a-very-secret-api-key, signature=, useEnterpriseCredentials=false, useSystemProxy=false, timeout=500}
> 2019-05-02T11:33:26,317 | ERROR | pool-11-thread-1 | DefaultGeolocationResolver       | 237 - org.opennms.features.geolocation.services - 24.1.0.SNAPSHOT | Error resolving address 'Schwieberdinger Str. 5, Zuffenhausen, 70435, Germany': timeout
> java.net.SocketTimeoutException: timeout
> 	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.java:662) ~[?:?]
> 	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.java:670) ~[?:?]
> 	at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.java:154) ~[?:?]
> 	at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.java:136) ~[?:?]
> 	at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115) ~[?:?]
> 	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[?:?]
> 	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[?:?]
> 	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[?:?]
> 	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[?:?]
> 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[?:?]
> 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[?:?]
> 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:221) ~[?:?]
> 	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:172) ~[?:?]
> 	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) ~[?:?]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
> 	at java.lang.Thread.run(Thread.java:834) [?:?]
> ```
> 
> the second try was successful
> 
> ```
> 2019-05-02T11:33:33,599 | DEBUG | pool-11-thread-1 | DefaultGeolocationResolver       | 237 - org.opennms.features.geolocation.services - 24.1.0.SNAPSHOT | Successfully resolved address 'Schwieberdinger Str. 5, Zuffenhausen, 70435, Germany': Active Geocoder with id 'google' resolved to long/lat: 9.1681678/48.8258935
> ```
> 
> After that all works fine.

I tested all of the geocoders and they worked fine. There was no timeout. Did this occur only once or was this seen more often? 
Side note: The Google Geocoder is the only one using a default try by default. The other's just fail.

**Comment by mvrueden** [comment]
I addressed all requested changes/feedback. I am done from my point of view, if I am not told otherwise :)

**Comment by mvrueden** [comment]
> Looks like this leaves the configuration options in `opennms.properties` intact; if that configuration is no longer valid, it should probably be removed in this PR as well?
> 
> https://github.com/OpenNMS/opennms/blob/develop/opennms-base-assembly/src/main/filtered/etc/opennms.properties#L535-L564, plus a few other sections, mention Geocoding there right now.

Done!

**Pull request merged**